### ISSUE TITLE
Use measures that are not of the form `f(y, yhat)` but `f(fitresult)`

### ISSUE THREAD

**Issue by dpaetzel** [JuliaAI/MLJTuning.jl#202]
Hi, thank you for developing and maintaining this part of MLJ! :slightly_smiling_face: 

I was wondering how one would go about the following:

I created a custom MLJ model type (let's call it `CustomModel`) which uses internally a custom optimizer (actually, an Evolutionary Algorithm but this is not important) with a custom objective function `cost`[^1]. I'd now like to perform hyperparameter optimization with respect to `cost`. At that, `cost` is computed by the inner optimizer anyways (since it's optimizing for it) and I can make the `fitresult`s of `CustomModel` contain the highest achieved `cost` value by the inner optimizer. What I'd like to do is provide instead of e.g. `measure=mae` something like `measure=custom` where `custom` accepts `fitresult`s (or something along those lines).

I looked mostly into [`tuned_models.jl`](https://github.com/JuliaAI/MLJTuning.jl/blob/dev/src/tuned_models.jl) and [`resampling.jl`](https://github.com/JuliaAI/MLJBase.jl/blob/dev/src/resampling.jl#L1243) and (presumably since I'm not familiar enough with the code) I only saw ways to achieve this that look like a lot of work. Maybe there is another way?

So far, it looks to me like writing my own version of `TunedModel` would be less work than changing the existing code. Maybe you can give me a hint as to what I'm missing? Did no one run into this so far?

Thank you for your time!

[^1]: Let's assume that `cost` is actually a decent measure for the inner optimizer. For example, `cost` does not only include predictive performance but also model complexity and it's therefore not of the form `cost(y, yhat)` but more like `cost(y, yhat, complexity, otherstuff)`.

**Comment by ablaom** [comment]
It seems likely `TunedModel` is a stretch for your purpose. I'm a little confused because you say you are already doing optimization internally - so what then is the role of MLJTuning.jl? 

There is the possibility of using a custom [selection heuristic](https://github.com/JuliaAI/MLJTuning.jl#how-do-i-implement-a-new-selection-heuristic). By default, `Tuned Model` just picks out the model (hyper-parameter set) with the lowest loss / highest score, but a custom selection heuristic allows for something different (e.g., "parsimonious" selection) so long as you can make a decision based on just the model evaluation history. Currently, what it written to the history is  not controlled by the user (some of it depends on the tuning strategy, such as `Grid`, `RandomSearch`, etc) but perhaps we could add a `TunedModel` hyperparameter `f` that gets applied to each `(model, fitresult)` (eg. returning model complexity) and we could arrange to have that written to the history in all cases. Just brainstorming here...

**Comment by dpaetzel** [comment]
Thanks for responding!

> It seems likely TunedModel is a stretch for your purpose. I'm a little confused because you say you are already doing optimization internally - so what then is the role of MLJTuning.jl?

Sorry, I may have not been clear enough there. I meant _internal optimization_ in the sense of candidate model fitting (i.e. model parameter optimization; e.g. fitting an NN's parameters to the data using backprop, just in my case an Evolutionary Algorithm is used instead of gradient descent).

> There is the possibility of using a custom selection heuristic.

Thank you, I had somehow missed that.

As I understand it, the selection heuristic `best(…)` is applied only at the last step. This means that for strategies like `Grid`, `RandomSearch`, `LatinHypercube` (where the next search point is independent of the search history so far), if I was able to write to the history additional things (like you proposed) this could work; I'd simply set `measure` to an arbitrary thing (e.g. `mae`).

However, for more sophisticated strategies like `TreeParzen` this is probably not enough because, as I understand it, each next search point is selected based on the history of hyperparametrizations and `measure` values.

I myself will probably for now stick with the simpler `Grid`/`LatinHypercube`-based approaches anyways so the latter wouldn't be a nuissance for me at the moment.

I'll look into this a bit more and then comment what I find out.

**Comment by dpaetzel** [comment]
I was able to adjust [`tuned_models.jl`](https://github.com/JuliaAI/MLJTuning.jl/blob/dev/src/tuned_models.jl) such that the following code works (`myextra` only returns the tree field for simplicity's sake)

```julia
using MLJ
DTRegressor = @load DecisionTreeRegressor pkg = DecisionTree verbosity = 0

N = 300
X, y = rand(N, 3), rand(N)
X = MLJ.table(X)

model = DTRegressor()

space = [
    range(model, :max_depth; lower=1, upper=5),
    range(
        model,
        :min_samples_split;
        lower=ceil(0.001 * N),
        upper=ceil(0.05 * N),
    ),
]

function myextra(model, fparams)
    # fparams = fitted_params(fitted_params(resampling_machine).machine)
    return fparams.tree
end

modelt = TunedModel(;
    model=model,
    resampling=CV(; nfolds=3),
    tuning=LatinHypercube(; gens=30),
    range=space,
    measure=mae,
    n=2,
    userextras=myextra,
)

macht = machine(modelt, X, y)
MLJ.fit!(macht; verbosity=1000)
display(report(macht).history[1].userextras)
```

However, I now have the problem that this only yields a single evaluation of `myextra` despite using 3-fold CV. What I would want is evaluating `myextra` once for each fold but the CV machinery (i.e. `evaluate`) is in `MLJBase`.

I guess I'll have to pass `userextras` to `evaluate` as well and then alter the `PerformanceEvaluation` struct (instances of which `evaluate` returns) … Or is there a way to have additional functions be evaluated by `evaluate` that I'm not seeing?

Are you interested

**Comment by dpaetzel** [comment]
My bad, I just noticed right after clicking comment that there is `PerformanceEvaluaten.fitted_params_per_fold` that I can use.

**Comment by ablaom** [comment]
I'd be happy to review such a PR.

Don't have a great idea for the name. What about `history_additions`?

I know I said `history_additions(model, fitresult)` for the signature but `fitresult` is not really public API. Rather, can we do `history_additions(model, fp)` where `fp` is something of the form `fitted_params(model, fitresult)` (equivalently, `fitted_params(mach)`, where `mach = machine(model, data...) |> fit!`)?

**Comment by ablaom** [comment]
I think this issue is resolved, albeit through a different means than originally posed. 

Closing for now.
